packages/explorer/src/app/(explorer)/[chainName]/worlds/[worldAddress]/decode/DecodeForm.tsx

```diff
@@ -43,12 +56,49 @@ function isAbiError(item: AbiItem): item is AbiItem & { type: "error" } {
 export function DecodeForm() {
   const { data: worldData, isLoading: isWorldAbiLoading } = useWorldAbiQuery();
   const { data: systemData, isLoading: isSystemAbisLoading } = useSystemAbisQuery();
-  const [abiItem, setAbiItem] = useState<AbiFunction | AbiError>();
+  const [decoded, setDecoded] = useState<AbiFunction | AbiError | ResourceItem | SelectorDatabase>();
+  const [encoded, setEncoded] = useState<string | undefined>();
   const form = useForm<z.infer<typeof formSchema>>({
     resolver: zodResolver(formSchema),
   });
+  const results = useMemo<Result[]>(() => {
+    if (!decoded) return [];
+    if (decoded.type === "resource") {
+      return [{ type: decoded.resource.type, label: resourceToLabel(decoded.resource) }];
+    }
+    if (decoded.type === "function" || decoded.type === "error") {
+      return [{ type: decoded.type, label: formatAbiItem(decoded) }];
+    }
+    if (decoded.type === "signature") {
+      return decoded.selectors.map((selector) => ({ type: "signature", label: selector }));
+    }
+    return [];
+  }, [decoded]);
+  const decodedFunctionCalls = useMemo(() => {
+    try {
+      if (!decoded || !encoded) return;
+      if (decoded.type === "function") {
+        return [{ ...decodeFunctionData({ abi: [decoded], data: encoded as Hex }), abi: decoded }];
+      }
+      if (decoded.type === "error") {
+        return [{ ...decodeErrorResult({ abi: [decoded], data: encoded as Hex }), abi: decoded }];
+      }
+      if (decoded.type === "signature") {
+        return decoded.selectors.map((selector) => {
+          const abiItem = parseAbiItem(`function ${selector}`) as AbiFunction;
+          return { ...decodeFunctionData({ abi: [abiItem], data: encoded as Hex }), abi: abiItem };
+        });
+      }
+    } catch (error) {
+      console.error("Error decoding function data", error);
+      // ignore error
+    }
+  }, [decoded, encoded]);
 
-  function onSubmit({ selector }: z.infer<typeof formSchema>) {
+  async function onSubmit({ encodedData }: z.infer<typeof formSchema>) {
+    setEncoded(encodedData);
+    if (!encodedData) return;
```

Comment:
should this return a line earlier?

Reply:
thought it should reset the `encodedData` if the form was submitted with empty data

Reply:
Ah right, makes sense